onnxruntime/core/graph/graph.cc

```diff
@@ -4046,6 +4070,210 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
+                                                       NodeAttributes& attributes) {
+  for (auto& attribute : attributes) {
+    auto& attr_proto = attribute.second;
+    if (attr_proto.has_g()) {
+      RenameSubgraphDependentNames(name_mapping, *attr_proto.mutable_g());
+    }
+
+    for (auto& subgraph : *attr_proto.mutable_graphs()) {
+      RenameSubgraphDependentNames(name_mapping, subgraph);
+    }
+  }
+}
+
+Status Graph::InlineIfSubgraph(const Graph& graph_to_inline, Node& if_node) {
+  std::string unique_id{if_node.Name()};
+  unique_id.append("_if_").append(graph_to_inline.Name());
+
+  unique_id = GenerateNodeName(unique_id);
+
+  auto make_unique = [&unique_id](const std::string& name) {
+    return unique_id + '_' + name;
+  };
+
+  // Check if the name is an input or implicit input.
+  // These are not renamed.
+  InlinedHashSet<std::string_view> if_all_inputs;
+  const auto if_inputs = if_node.InputDefs();
+  const auto if_implicit_inputs = if_node.ImplicitInputDefs();
+  if_all_inputs.reserve(if_inputs.size() + if_implicit_inputs.size());
+  for (const auto* input : if_inputs) {
+    const auto& name = input->Name();
+    ORT_IGNORE_RETURN_VALUE(if_all_inputs.emplace(name));
+  }
+
+  for (const auto* input : if_implicit_inputs) {
+    const auto& name = input->Name();
+    ORT_IGNORE_RETURN_VALUE(if_all_inputs.emplace(name));
+  }
+
+  // Name mapping from the graph to inline to the graph we are inlining into
+  // we also use this to process any subgraphs in the graph we are inlining
+  InlinedHashMap<std::string, std::string> name_mapping;
+
+  // We are going to map the outputs of the graph to inline to the outputs of the If node.
+  // They are assumed to be in the same order.
+  const auto node_output_defs = if_node.OutputDefs();
+  const auto graph_output_defs = graph_to_inline.GetOutputs();
+  ORT_ENFORCE(node_output_defs.size() >= graph_output_defs.size());
+  for (size_t i = 0; i < graph_output_defs.size(); ++i) {
+    name_mapping.emplace(graph_output_defs[i]->Name(), node_output_defs[i]->Name());
+  }
+
+  // Process constant nodes and initializers first
+  // to create defs from them to retain the type, and make renaming easier
+  for (const auto& node : graph_to_inline.Nodes()) {
+    if (node.OpType() == kConstant) {
+      // Copy constant nodes _value to name_to_initial_tensor_
+      ONNX_NAMESPACE::NodeProto constant_node_proto{};
+      node.ToProto(constant_node_proto);
+
+      const auto& constant_name = constant_node_proto.output(0);
+      // Check if this is an output of the graph
+      std::string new_name;
+      auto hit = name_mapping.find(constant_name);
+      if (hit != name_mapping.cend()) {
+        new_name = hit->second;
+      } else {
+        new_name = GenerateNodeArgName(make_unique(constant_name));
+        name_mapping.emplace(constant_name, new_name);
+      }
+      ORT_RETURN_IF_ERROR(AddConstantProtoAsInitializer(constant_node_proto, new_name));
+    }
```

Comment:
IIRC the onnxruntime::Graph ctor converts all Constant nodes to initializers, so given this is a Graph instance I wouldn't have expected any Constant nodes to remain. Different story if you were inlining the GraphProto in the then/else attribute though.